[Multiplayer] Add User panels

Naeferith opened the pull request:
![User Panels](http://i.imgur.com/G4zGuU4.png)
_flyte's design_

Naeferith committed:
Test Case for User panels

Naeferith committed:
Room's User panels

RemieRichards commented on osu.Game/Screens/Multiplayer/MultiUserPanel.cs:
```diff
@@ -0,0 +1,215 @@
+                switch (state)
+                {
+                    case UserState.Guest:
+                        statusString = "Hosting a Room";
```
aren't these backwards?

Naeferith replied:
Indeed, thanks :)

RemieRichards commented on osu.Game/Screens/Multiplayer/MultiUserPanel.cs:
```diff
@@ -0,0 +1,215 @@
+        public const int STATUS_HEIGHT = 33;
+        public const float PANEL_WIDTH = 0.25f;
+
+        public void UpdatePanel(MultiUserPanel panel)
```
Is this supposed to do more than just set the spritetext? otherwise just make ```statusString``` a property that sets the spritetext

huoyaoyuan replied:
an instance method that always taking this as a parameter is stupid

turbedi commented on osu.Desktop.VisualTests/Tests/TestCaseMultiUserPanel.cs:
```diff
@@ -0,0 +1,57 @@
+        private MultiUserPanel test;
+        private FlowContainer panelContainer;
+        public override string Name => @"MultiUserPanel";
+        public override string Description => @"User pannel in a multiplayer room";
```
typo

RemieRichards commented on osu.Game/Screens/Multiplayer/MultiUserPanel.cs:
```diff
@@ -0,0 +1,215 @@
+                    Origin = Anchor.BottomCentre,
+                    Colour = new Color4(17, 136, 170, 128),
+                },
+                new Container //Avatar
```
Avatars are implemented, you should probably store a ```User``` on this class, and instantiate a new ```Avatar(some_user)```

huoyaoyuan replied:
There is an already done `Avatar`

huoyaoyuan commented on osu.Game/Screens/Multiplayer/MultiUserPanel.cs:
```diff
@@ -0,0 +1,215 @@
+            statusSprite.Text = statusString;
+        }
+
+        public MultiUserPanel(string user = "Unknown")
```
Take an instance of `User` class, instead of a string

huoyaoyuan commented on osu.Game/Screens/Multiplayer/MultiUserPanel.cs:
```diff
@@ -0,0 +1,215 @@
+                            Children = new Drawable[]
+                            {
+
+                                new Container //Country Flag
```
`DrawableFlag`

huoyaoyuan commented:
btw where do you get the design?

Naeferith commented:
I got it from @peppy himself

MrTheMake commented on osu.Desktop.VisualTests/Tests/TestCaseMultiUserPanel.cs:
```diff
@@ -0,0 +1,57 @@
+
+            AddButton(@"Switch status", () => action(0));
+
+            Add(panelContainer = new FlowContainer //Positionning container
```
typo in comment

Naeferith committed:
Update MultiUserPanel.cs

- Replace all FlowContainer with FillFlowContainer
- Replace a Container with Avatar
- Replace a Container with DrawableFlag
- Revert StatusSring in the switch
- Remove Updatepanel method
- Other modifications

Naeferith committed:
Update TestCaseMultiUserPanel.cs due to MultiUserPanel.cs modifications

huoyaoyuan commented on osu.Desktop.VisualTests/Tests/TestCaseMultiUserPanel.cs:
```diff
@@ -0,0 +1,66 @@
+    {
+        private MultiUserPanel test;
+        private FillFlowContainer panelContainer;
+        public User User = new User
```
This could be created in `Reset()`, rather than save it(see visual test of leaderboard)

huoyaoyuan commented on osu.Desktop.VisualTests/Tests/TestCaseMultiUserPanel.cs:
```diff
@@ -0,0 +1,66 @@
+
+        private void action(int action)
+        {
+            switch (action)
```
What's this for?

huoyaoyuan commented on osu.Desktop.VisualTests/Tests/TestCaseMultiUserPanel.cs:
```diff
@@ -0,0 +1,66 @@
+
+            AddStep(@"Switch status", () => action(0));
+
+            Add(panelContainer = new FillFlowContainer
```
My suggestion:
  - make a generic container
  - allow adding new panel
  - set host via the container, not a reference to specified panel

huoyaoyuan commented on osu.Game/Screens/Multiplayer/MultiUserPanel.cs:
```diff
@@ -0,0 +1,195 @@
+        private OsuSpriteText userSprite;
+        private OsuSpriteText statusSprite;
+
+        public const int PANEL_HEIGHT = 100;
```
private consts

huoyaoyuan commented on osu.Game/Screens/Multiplayer/MultiUserPanel.cs:
```diff
@@ -0,0 +1,195 @@
+            }
+        }
+
+        private string userName;
```
consider if you DO need fields for username and status string

huoyaoyuan commented on osu.Game/Screens/Multiplayer/MultiUserPanel.cs:
```diff
@@ -0,0 +1,195 @@
+                new Box
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Colour = new Color4(34,34,34, 255),
```
format code manually(the IDE formatter doesn't work in nested object initializer lol)

Naeferith committed:
Update MultiUserPanel.cs

- Consts are private
- Formated code
- Revmove userName and statusString

Naeferith committed:
Change default state on panel creation

Host -> Guest

Naeferith committed:
Step buttons modifications

- Add a "Add panel" step
- Change "Switch status" to "Random Host"

Naeferith committed:
Remove host reference + Add methods

-Split the action() method into separated methods

smoogipooo commented on osu.Game/Screens/Multiplayer/MultiUserPanel.cs:
```diff
@@ -0,0 +1,191 @@
+                    case UserState.Guest:
+                        statusSprite.Text = "Waiting in a Room";
+                        break;
+
```
We don't use new lines in switch statements.

smoogipooo commented on osu.Game/Screens/Multiplayer/MultiUserPanel.cs:
```diff
@@ -0,0 +1,191 @@
+        private const int STATUS_HEIGHT = 33;
+        private const float PANEL_WIDTH = 0.25f;
+
+
```
Extra new line.

smoogipooo commented on osu.Game/Screens/Multiplayer/MultiUserPanel.cs:
```diff
@@ -0,0 +1,191 @@
+                switch (state)
+                {
+                    case UserState.Guest:
+                        statusSprite.Text = "Waiting in a Room";
```
"Waiting in a room" // "Hosting a room" (casing)

smoogipooo commented on osu.Game/Screens/Multiplayer/MultiUserPanel.cs:
```diff
@@ -0,0 +1,191 @@
+                    Anchor = Anchor.TopRight,
+                    Origin = Anchor.TopRight,
+                    Direction = FillDirection.Full,
+
```
No spaces in object initializers.

smoogipooo commented on osu.Game/Screens/Multiplayer/MultiUserPanel.cs:
```diff
@@ -0,0 +1,191 @@
+                    Height = STATUS_HEIGHT,
+                    Anchor = Anchor.BottomCentre,
+                    Origin = Anchor.BottomCentre,
+                    Colour = new Color4(17, 136, 170, 128),
```
Use colours from the OsuColour palette.
```
[BackgroundDependencyLoader]
private void load(OsuColour colours)
{
   myBox.Colour = colours.MyColour;
}
```

smoogipooo commented on osu.Game/Screens/Multiplayer/MultiUserPanel.cs:
```diff
@@ -0,0 +1,191 @@
+                {
+                    Anchor = Anchor.BottomLeft,
+                    Origin = Anchor.BottomLeft,
+                    Text = "Waiting in a Room",
```
Also casing here. Consider making these text pieces into private consts for now (will be localised later).

smoogipooo commented on osu.Desktop.VisualTests/Tests/TestCaseMultiUserPanel.cs:
```diff
@@ -0,0 +1,76 @@
+
+        private void rndHost()
+        {
+            int newHost = RNG.Next(panelContainer.Children.Count());
```
Let's avoiod rng in test cases - just make this cycle through them.

smoogipooo commented on osu.Desktop.VisualTests/Tests/TestCaseMultiUserPanel.cs:
```diff
@@ -0,0 +1,76 @@
+            {
+                panels.State = MultiUserPanel.UserState.Guest;
+            }
+            newHostPanel.State = MultiUserPanel.UserState.Host;
```
This is going to be a pretty common use case. I suggest making a MultiUserPanelContainer which has, say, a property such as
```
public User Host
{
    get { return Children.FirstOrDefault(c => c.Host); }
    set { Children.ForEach(c => c.Host = c == value); }
}
```

Naeferith replied:
The c parameter is defined as a Drawable and not a MultiUserPanel, is there a way I can cast this parameter ?

huoyaoyuan replied:
Use a generic container: `FillFlowContainer<MultiUserPanel>`

Naeferith replied:
The FirstOrDefault method will return a MultiUserPanel, should I change the type of ```Host``` into MultiUserPanel or store a User in the MultiUserPanel class so I can do 
```return Children.FirstOrDefault(c => c.Host).User;``` ?

Naeferith replied:
Setting ```Host``` as a ```MultiUserPanel``` should do the trick

huoyaoyuan replied:
If there is a type `MultiUserPanelContainer`, it can keep a reference to host panel

smoogipooo commented on osu.Game/Screens/Multiplayer/MultiUserPanel.cs:
```diff
@@ -0,0 +1,191 @@
+            };
+        }
+
+        public enum UserState
```
Move enum to separate file.

Naeferith replied:
You told me to use a bool instead right ?

smoogipooo commented on osu.Game/Screens/Multiplayer/MultiUserPanel.cs:
```diff
@@ -0,0 +1,191 @@
+    public class MultiUserPanel : ClickableContainer, IStateful<MultiUserPanel.UserState>
+    {
+        private UserState state;
+        public UserState State
```
Would avoid the enum for now and keep it simple with a `bool Host`.

smoogipoo reviewed (changes requested):
A few changes requested, but pretty okay otherwise!

I'm not sure about being too pent up on having Anchors/Origins at the top of object initializers, so I'll let it pass for now.

Naeferith committed:
Generic container for User panels

Naeferith committed:
Update MultiUserPanel.cs

- Use osu! colours
- Remove IStateful implementation
- Remove unnecessary new lines
- Put Raw Text into variables

Naeferith committed:
Use MultiUserPanelContainer + method modificaion

- Random Host -> Host Rotation

MrTheMake commented on osu.Game/Screens/Multiplayer/MultiUserPanel.cs:
```diff
@@ -0,0 +1,192 @@
+                    return;
+
+                host = value;
+                switch(host)
```
I would use an if-statement here since the argument is a boolean

Naeferith committed:
Replace switch with If-statement

Naeferith committed:
Remove unused using

Naeferith committed:
Remove unused usings

Naeferith committed:
Update osu.Desktop.VisualTests.csproj

Naeferith commented:
See #825
